README.md
# R.A.T — Reginald's Aim Trainer 🖱️

[▶️ Play](https://ven1xxy.github.io/aim-trainer) ● [📜 Changelog](CHANGELOG.md)

An aim trainer that runs in your web browser.

`⚠️ R.A.T is in alpha. Bugs are to be expected.`


CHANGELOG.md
# R.A.T — Changelog 📜

[⬅️ Back to README](README.md)

## Alpha

### `rat.ver.a.01`

- Better mobile support (still not good tho).
- Improved visual quality.
- High DPI support.
- Added menus but they're empty for now.
- Updated difficulties, removed ones may become different game modes.
- Hover effects for input elements.
- Better canvas sizing / Resize handling.
- Cursor changed to crosshair.
- A little bit of refactoring.
- Created README.md and CHANGELOG.md


js/trainer.js
var gameOptions = gameModes["standard"];

var canvas;
var context;

var targets;

/* 
2px of tolerance, so hitboxes are actually 2px larger on every side compared to the texture
we do this to sort of compensate for dumbass cursors.
*/
var lenience = 2;

var targetImage = new Image();
targetImage.src = "./assets/images/target.png";

var hitAudio = new Audio("./assets/sounds/hit.wav");
var missAudio = new Audio("./assets/sounds/miss.wav");
var interactionAudio = new Audio("./assets/sounds/interaction.wav");
var sliderAudio = new Audio("./assets/sounds/slider.wav");

var backgroundColor = "#262335";

window.onload = function (e) {
  canvas = document.getElementsByTagName("canvas")[0];
  context = canvas.getContext("2d");

  canvas.onclick = function (e) {
    var rect = e.target.getBoundingClientRect();
    var x = e.clientX - rect.left; // x position within the element.
    var y = e.clientY - rect.top; // y position within the element.

    handleClick(x, y);
  };

  var modeSelector = document.getElementsByTagName("select")[0];

  modeSelector.onclick = function (e) {
    playSound(interactionAudio);
  };

  modeSelector.onchange = function (e) {
    gameOptions = gameModes[modeSelector.value];
    playSound(sliderAudio);
    beginGame();
  };

  beginGame();
  draw();
};

function populate() {
  for (let i = 0; i < gameOptions.targetCount; i++) {
    targets.push(createRandomTarget());
  }
}

function handleClick(x, y) {
  var newTargets = [];

  var hit = false;

  for (let i = 0; i < targets.length; i++) {
    const target = targets[i];

    var drawTarget = target.calculateDrawTarget();

    if (
      hit == false &&
      x >= drawTarget.x - lenience &&
      x <= drawTarget.x + drawTarget.width + lenience &&
      y >= drawTarget.y - lenience &&
      y <= drawTarget.y + drawTarget.height + lenience
    ) {
      playSound(hitAudio);
      hit = true;
      newTargets.push(createRandomTarget());
      continue;
    }

    newTargets.push(target);
  }

  if (!hit) {
    playSound(missAudio);
  }

  targets = newTargets;
}

var lastDraw;

function draw() {
  var now = Date.now();
  var deltaTime = (now - lastDraw) / 1000;

  context.clearRect(0, 0, canvas.width, canvas.height);
  context.fillStyle = backgroundColor;
  context.fillRect(0, 0, canvas.width, canvas.height);
  context.fillStyle = "#ffffff";

  var newTargets = [];

  for (let i = 0; i < targets.length; i++) {
    drawTarget(targets[i]);

    targets[i].scale -= gameOptions.targetScaleSpeed * deltaTime;

    if (targets[i].scale > 0) {
      newTargets.push(targets[i]);
    } else {
      newTargets.push(createRandomTarget());
      playSound(missAudio);
    }
  }

  targets = newTargets;

  lastDraw = now;
  requestAnimationFrame(draw);
}

function playSound(audio) {
  if (audio.readyState == HTMLMediaElement.HAVE_ENOUGH_DATA) {
    var clone = audio.cloneNode();
    clone.currentTime = 0;
    clone.play();
  }
}

function drawTarget(target) {
  var drawTarget = target.calculateDrawTarget();
  context.drawImage(
    targetImage,
    drawTarget.x,
    drawTarget.y,
    drawTarget.width,
    drawTarget.height
  );
}

function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min) + min); // The maximum is exclusive and the minimum is inclusive
}

function createRandomTarget(depth = 0) {
  var target = new Target(
    getRandomInt(0, canvas.width - 100),
    getRandomInt(0, canvas.height - 100),
    gameOptions.targetWidth,
    gameOptions.targetHeight,
    1
  );

  //  Avoid spawning targets inside each other.
  if (depth < 5) {
    // stop trying after 5 retries
    targets.forEach((compareTarget) => {
      if (
        target.x >= compareTarget.x - compareTarget.width &&
        target.x <=
          compareTarget.x + compareTarget.width + compareTarget.width &&
        target.y >= compareTarget.y - compareTarget.height &&
        target.y <=
          compareTarget.y + compareTarget.height + compareTarget.height
      ) {
        console.log("collision prevented.");
        return createRandomTarget(depth + 1);
      }
    });
  } else {
    console.log("couldn't prevent collision. (too many retries)");
  }

  return target;
}

function beginGame() {
  targets = [];

  populate();
}


js/renderer.js
function initRenderer() {
  resizeCanvas();
  context.imageSmoothingEnabled = true;
  context.imageSmoothingQuality = "high";
  context.textBaseline = "top";
}

function drawTarget(target, texture) {
  var drawTarget = target.calculateDrawTarget();
  context.drawImage(
    texture,
    drawTarget.x,
    drawTarget.y,
    drawTarget.width,
    drawTarget.height
  );
}

function resizeCanvas() {
  var oldW = canvas.width;
  var oldH = canvas.height;
  canvas.width = document.body.clientWidth * 0.7;
  canvas.height = document.body.clientHeight * 0.8;

  rect = canvas.getBoundingClientRect(); // Update rect
  handleDPR();

  var multX = canvas.width / oldW;
  var multY = canvas.height / oldH;

  console.log(
    `canvas scaled: x: ${oldW} → ${canvas.width} (${multX}x) // y: ${oldH} → ${canvas.height} (${multY}x)`
  );

  for (let i = 0; i < targets.length; i++) {
    targets[i].x *= multX;
    targets[i].y *= multY;
  }
}

function handleDPR() {
  dpr = (window.devicePixelRatio || 1) * sampling;
  canvas.width = rect.width * dpr;
  canvas.height = rect.height * dpr;
  var ctx = canvas.getContext("2d");
  ctx.scale(dpr, dpr);
  rect = canvas.getBoundingClientRect(); // Update rect
}

function clearCanvas() {
  context.fillStyle = backgroundColor;
  context.fillRect(0, 0, canvas.width, canvas.height);
  context.fillStyle = "#ffffff";
  /* TODO: Try optimizing this, using seperate layers for background & game,
  also grid background */
}

function drawFPS(frameTime) {
  var fps = Math.round(1 / frameTime);
  context.fillText(`${fps} FPS (${frameTime * 1000} ms)`, 16, 16);
}


js/classic.js
var difficulties = {
  unlosable: {
    targetCount: 3,
    targetWidth: 50,
    targetHeight: 50,
    targetScaleSpeed: -0.5,
  },

  easy: {
    targetCount: 1,
    targetWidth: 100,
    targetHeight: 100,
    targetScaleSpeed: 0.5,
  },

  medium: {
    targetCount: 4,
    targetWidth: 50,
    targetHeight: 50,
    targetScaleSpeed: 0.1,
  },

  hard: {
    targetCount: 4,
    targetWidth: 45,
    targetHeight: 45,
    targetScaleSpeed: 0.2,
  },

  nightmare: {
    targetCount: 4,
    targetWidth: 35,
    targetHeight: 35,
    targetScaleSpeed: 0.25,
  },

  dead: {
    targetCount: 2,
    targetWidth: 25,
    targetHeight: 25,
    targetScaleSpeed: 0.4,
  },
};

var gameOptions = difficulties["medium"];

var dpr = window.devicePixelRatio || 1; // Used for DPI Scaling.

var sampling = 1;

var canvas;
var context;
var rect;

var targets = [];

var lastDraw = Date.now();

/* 
2px of tolerance, so hitboxes are actually 2px larger on every side compared to the texture
we do this to sort of compensate for dumbass cursors.
*/
var lenience = 2;

var targetImage = new Image();
targetImage.src = "./assets/images/target.png";

var hitAudio = new Audio("./assets/sounds/hit.wav");
var missAudio = new Audio("./assets/sounds/miss.wav");
var interactionAudio = new Audio("./assets/sounds/interaction.wav");
var sliderAudio = new Audio("./assets/sounds/slider.wav");

var backgroundColor = "#262335";

window.onload = function (e) {
  /*  "No super" modals */
  var modals = document.getElementsByClassName("modal");
  for (let i = 0; i < modals.length; i++) {
    const modal = modals[i];
    modal.onclick = function (e) {
      noSuper(e);
    };
  }

  canvas = document.getElementsByTagName("canvas")[0];
  context = canvas.getContext("2d", { alpha: false });

  // Initialize renderer.
  initRenderer();

  canvas.onmousedown = function (e) {
    var x = e.clientX - rect.left; // x position within the element.
    var y = e.clientY - rect.top; // y position within the element.

    handleClick(x, y);
  };

  window.onresize = function (e) {
    resizeCanvas();
  };

  window.onscroll = function (e) {
    // Recalculate rect.
    rect = canvas.getBoundingClientRect();
  };

  var modeSelector = document.getElementsByTagName("select")[0];

  modeSelector.onclick = function (e) {
    playSound(interactionAudio);
  };

  modeSelector.onchange = function (e) {
    gameOptions = difficulties[modeSelector.value];
    playSound(sliderAudio);
    beginGame();
  };

  beginGame();
  draw();
};

function populate() {
  for (let i = 0; i < gameOptions.targetCount; i++) {
    targets.push(createRandomTarget(0));
  }
}

function handleClick(x, y) {
  var newTargets = [];

  var hit = false;

  // Reversed, so the target on top is checked first.
  for (let i = targets.length - 1; i >= 0; i--) {
    const target = targets[i];

    var drawTarget = target.calculateDrawTarget();

    if (
      hit == false &&
      x >= drawTarget.x - lenience &&
      x <= drawTarget.x + drawTarget.width + lenience &&
      y >= drawTarget.y - lenience &&
      y <= drawTarget.y + drawTarget.height + lenience
    ) {
      playSound(hitAudio);
      hit = true;
      newTargets[i] = createRandomTarget(0);
      continue;
    }

    newTargets[i] = target;
  }

  if (!hit) {
    playSound(missAudio);
  }

  targets = newTargets;
}

function draw() {
  var now = Date.now();
  var deltaTime = (now - lastDraw) / 1000;

  clearCanvas();

  drawFPS(deltaTime);

  var newTargets = [];

  for (let i = 0; i < targets.length; i++) {
    drawTarget(targets[i], targetImage);

    targets[i].scale -= gameOptions.targetScaleSpeed * deltaTime;

    if (targets[i].scale > 0) {
      newTargets.push(targets[i]);
    } else {
      newTargets.push(createRandomTarget(0));
      playSound(missAudio);
    }
  }

  targets = newTargets;

  lastDraw = now;
  requestAnimationFrame(draw);
}

function playSound(audio) {
  if (audio.readyState == HTMLMediaElement.HAVE_ENOUGH_DATA) {
    var clone = audio.cloneNode();
    clone.currentTime = 0;
    clone.play();
  }
}

function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min) + min); // The maximum is exclusive and the minimum is inclusive
}

function createRandomTarget(depth = 0) {
  var target = new Target(
    getRandomInt(0, canvas.width / dpr - 100),
    getRandomInt(0, canvas.height / dpr - 100),
    gameOptions.targetWidth,
    gameOptions.targetHeight,
    1
  );

  //  Avoid spawning targets inside each other.
  if (depth < 5) {
    // stop trying after 5 retries
    for (let i = 0; i < targets.length; i++) {
      const compareTarget = targets[i];
      if (
        target.x >= compareTarget.x - compareTarget.width &&
        target.x <= compareTarget.x + compareTarget.width &&
        target.y >= compareTarget.y - compareTarget.height &&
        target.y <= compareTarget.y + compareTarget.height
      ) {
        return createRandomTarget(depth + 1);
      }
    }
  } else {
    console.log("couldn't prevent collision. (too many retries)");
  }

  return target;
}

function beginGame() {
  targets = [];

  populate();
}

function showModal(modalID) {
  var modalsContainer = document.getElementById("modals");
  modalsContainer.className = "";
  var graphicsModal = document.getElementById(modalID);
  graphicsModal.className = "modal";
}

function hideModals() {
  // Hides *all* modals.
  var modalsContainer = document.getElementById("modals");
  modalsContainer.className = "hide";
  var modals = document.getElementsByClassName("modal");
  for (let i = 0; i < modals.length; i++) {
    const modal = modals[i];
    modal.className = "modal hide";
  }
}

function noSuper(e) {
  e.stopPropagation();
}
